src/main/java/slimeknights/tconstruct/smeltery/tileentity/CastingTileEntity.java

```diff
@@ -0,0 +1,236 @@
+    if (tank.getFluidAmount() > 0) {
+      return;
+    }
+
+    // completely empty -> insert current item into input
+    if (!isStackInSlot(0) && !isStackInSlot(1)) {
+      ItemStack stack = player.inventory.decrStackSize(player.inventory.currentItem, stackSizeLimit);
+      setInventorySlotContents(INPUT, stack);
+    }
+    // take item out
+    else {
+      // take out stack 1 if something is there, 0 otherwise
+      int slot = isStackInSlot(OUTPUT) ? OUTPUT : INPUT;
+
+      // Additional info: Only 1 item can be put into the casting block usually, however recipes
+      // can have ItemStacks with stacksize > 1 as output
+      // we therefore spill the whole contents on extraction.
+      ItemStack stack = getStackInSlot(slot);
+      if (slot == OUTPUT) {
+        // fire player smelt event?
+      }
+      ItemHandlerHelper.giveItemToPlayer(player, stack);
+      setInventorySlotContents(slot, ItemStack.EMPTY);
+
+      // send a block update for the comparator, needs to be done after the stack is removed
+      if (slot == OUTPUT) {
+        this.getWorld().notifyNeighborsOfStateChange(this.pos, this.getBlockState().getBlock());
+      }
+    }
+  }
+
+
+  @Override
+  @Nonnull
+  public int[] getSlotsForFace(Direction side) {
+    return new int[]{INPUT, OUTPUT};
+  }
+
+  @Override
+  public boolean canInsertItem(int index, ItemStack itemStackIn, @Nullable Direction direction) {
+    return index == INPUT && !isStackInSlot(OUTPUT);
+  }
+
+  @Override
+  public boolean canExtractItem(int index, ItemStack stack, Direction direction) {
+    return index == 1;
+  }
+
+  @Override
+  public void tick() {
+    // no recipe
+    if (recipe == null) {
+      return;
+    }
+    // fully filled
+    if (tank.getFluidAmount() == tank.getCapacity() && !tank.getFluid().isEmpty()) {
+      timer++;
+      if (!getWorld().isRemote) {
+        if (timer >= recipe.getCoolingTime()) {
+          if (recipe.consumesCast()) {
+            setInventorySlotContents(INPUT, ItemStack.EMPTY);
+          }
+          if (recipe.switchSlots()) {
+            setInventorySlotContents(INPUT, recipe.getRecipeOutput());
+          }
+          else {
+            setInventorySlotContents(OUTPUT, recipe.getRecipeOutput());
+          }
+          getWorld().playSound(null, pos, SoundEvents.BLOCK_LAVA_EXTINGUISH, SoundCategory.AMBIENT, 0.07f, 4f);
+
+          reset();
+
+          getWorld().notifyNeighborsOfStateChange(this.pos, this.getBlockState().getBlock());
+        }
+      }
+      else if (getWorld().rand.nextFloat() > 0.9f) {
+        world.addParticle(ParticleTypes.SMOKE, pos.getX() + getWorld().rand.nextDouble(), pos.getY() + 1.1d, pos.getZ() + getWorld().rand.nextDouble(), 0.0D, 0.0D, 0.0D);
+      }
+    }
+  }
+
+  protected AbstractCastingRecipe findRecipe(Fluid fluid) {
+
+    recipes.clear();
+    recipes = world.getRecipeManager().getRecipes((IRecipeType<AbstractCastingRecipe>) recipeType, this, world);
```

Comment:
This seems inefficient, if you are going to store the list, store it once per type. Alo not a fan of that cast, can we find a way to not need it?

Reply:
I think this can be turned to a `getRecipe()` (note the lack of an 's') with the
change to AbstractCastingRecipe you suggested on irc, which seems to not
need the cast. Testing is required and I'm trying to be very careful to avoid
breaking what already works.